MinPlatformPkg/Test/Library/TestPointLib/StandaloneMmTestPointCommunication.c

```diff
@@ -0,0 +1,379 @@
+    DEBUG((DEBUG_ERROR, "StandaloneMmTestPointSmiHandlerGetDataByOffset: StandaloneMmTestPoint get data in SMRAM or overflow!\n"));
+    SmiHandlerTestPointParameterGetDataByOffset->Header.ReturnStatus = (UINT64)(INT64)(INTN)EFI_ACCESS_DENIED;
+    goto Done;
+  }
+  
+  DataSize = 0;
+  Status = GetAllStandaloneMmTestPointData (&DataSize, NULL);
+  if (Status != EFI_BUFFER_TOO_SMALL) {
+    SmiHandlerTestPointParameterGetDataByOffset->Header.ReturnStatus = (UINT64)(INT64)(INTN)EFI_NOT_FOUND;
+    goto Done;
+  }
+  Data = AllocatePool (DataSize);
+  if (Data == NULL) {
+    SmiHandlerTestPointParameterGetDataByOffset->Header.ReturnStatus = (UINT64)(INT64)(INTN)EFI_OUT_OF_RESOURCES;
+    goto Done;
+  }
+  Status = GetAllStandaloneMmTestPointData (&DataSize, Data);
+  if (EFI_ERROR(Status)) {
+    SmiHandlerTestPointParameterGetDataByOffset->Header.ReturnStatus = (UINT64)(INT64)(INTN)Status;
+    goto Done;
+  }
+
+  //
+  // The SpeculationBarrier() call here is to ensure the previous range/content
+  // checks for the CommBuffer have been completed before calling into
+  // SmiHandlerTestPointCopyData().
+  //
+  SpeculationBarrier ();
+  SmiHandlerTestPointCopyData (
+    Data,
+    DataSize,
+    (VOID *)(UINTN)SmiHandlerTestPointGetDataByOffset.DataBuffer,
+    &SmiHandlerTestPointGetDataByOffset.DataSize,
+    &SmiHandlerTestPointGetDataByOffset.DataOffset
+    );
+
+  CopyMem (
+    SmiHandlerTestPointParameterGetDataByOffset,
+    &SmiHandlerTestPointGetDataByOffset,
+    sizeof(SmiHandlerTestPointGetDataByOffset)
+    );
+
+  SmiHandlerTestPointParameterGetDataByOffset->Header.ReturnStatus = 0;
+
+Done:
+  if (Data != NULL) {
+    FreePool (Data);
+  }
+}
+
+/**
+  Dispatch function for a Software SMI handler.
+
+  Caution: This function may receive untrusted input.
+  Communicate buffer and buffer size are external input, so this function will do basic validation.
+
+  @param DispatchHandle  The unique handle assigned to this handler by SmiHandlerRegister().
+  @param Context         Points to an optional handler context which was specified when the
+                         handler was registered.
+  @param CommBuffer      A pointer to a collection of data in memory that will
+                         be conveyed from a non-StandlaoneMm environment into an StandaloneMm environment.
+  @param CommBufferSize  The size of the CommBuffer.
+
+  @retval EFI_SUCCESS Command is handled successfully.
+**/
+EFI_STATUS
+EFIAPI
+StandaloneMmTestPointSmiHandler (
+  IN EFI_HANDLE  DispatchHandle,
+  IN CONST VOID  *Context         OPTIONAL,
+  IN OUT VOID    *CommBuffer      OPTIONAL,
+  IN OUT UINTN   *CommBufferSize  OPTIONAL
+  )
+{
+  SMI_HANDLER_TEST_POINT_PARAMETER_HEADER     *SmiHandlerTestPointParameterHeader;
+  UINTN                                       TempCommBufferSize;
+
+  DEBUG((DEBUG_ERROR, "StandaloneMmTestPointSmiHandler Enter\n"));
```

Comment:
Is the message `DEBUG_ERROR` to get it to show in RELEASE builds?

A simple function entry message doesn't seem important enough to show in those builds if that's the case.

Reply:
Fixed.